src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,300 @@
+   * `matchNewObs(k, sz)` is called. For example, given:
+   *
+   * Obs[typeOf(c_z1)] = {(c_z1, (+ 1 1), null)}
+   * Pool[typeOf(c_z1)] = {(c_+ c_z2 c_z3)}
+   * CandSols = {}
+   *
+   * Then calling `matchNewObs(c_z1, (c_+ c_z2 c_z3))` will result in:
+   *
+   * Obs[typeOf(c_z1)] = {(c_z1, (+ 1 1), null), (c_z4, 1, null)}
+   * Pool[typeOf(c_z1)] = {(c_+ c_z2 c_z3)}
+   * CandSols = {c_z1 -> {(c_+ c_z4 c_z4)}}
+   *
+   * and will return `{typeOf(c_z4) -> {c_z4}}`.
+   *
+   * Notice that `c_z2` and `c_z3` are not returned as new sub-obligations.
+   * Instead, `(c_+ c_z2 c_z3)` is instantiated with a new skolem `c_z4`, which
+   * is then added to the set of obligations. This is done to allow the reuse of
+   * patterns in `Pool`. Also, notice that only one new skolem/sub-obligation is
+   * generated. That's because the builtin analogs of `c_z2` and `c_z3` match
+   * with the same builtin term `1`.
+   *
+   * @param ob free var whose builtin term we need to match
+   * @param sz a pattern to match `ob`s builtin term with
+   * @return a set of new obligations to satisfy if the match succeeds
+   */
+  TypeObligationSetMap matchNewObs(Node ob, Node sz);
+
+  /** mark obligation `k` as solved
+   *
+   * This function first marks `s` as the complete/constant solution for
+   * `ob`. Then it substitutes all instances of `ob` in partial solutions to
+   * other obligations with `s`. The function repeats those two steps for each
+   * partial solution that gets completed because of the second step. For
+   * example, given:
+   *
+   * CandSols = {
+   *  mainOb -> {(+ z1 1)},
+   *  z1 -> {(* z2 x)},
+   *  z2 -> {2}
+   * }
+   * Sol = {z2 -> 2}
+   *
+   * Then calling `markSolved(z2, 2)` will result in:
+   *
+   * CandSols = {
+   *  mainOb -> {(+ z1 1), (+ (* 2 x) 1)},
+   *  z1 -> {(* z2 x), (* 2 x)},
+   *  z2 -> {2}
+   * }
+   * Sol = {mainOb -> (+ (* 2 x) 1), z1 -> (* 2 x), z2 -> 2}
+   *
+   * Note: example uses builtin terms instead of sygus terms for simplicity.
+   *
+   * @param ob free var to mark as solved and substitute
+   * @param sol constant solution to `ob`
+   */
+  void markSolved(Node k, Node s);
+
+  /**
+   * Initialize a sygus enumerator and a candidate rewrite database for each
+   * sygus datatype type.
+   *
+   * @param stn The sygus datatype type encoding the syntax restrictions
+   */
+  void initialize(TypeNode stn);
+
+  /**
+   * Remove solved obligations from the given set of obligations.
+   *
+   * @param unsolvedObs A set of obligations containing solved ones
+   */
+  void removeSolvedObs(TypeObligationSetMap& obs);
+
+  /**
+   * Replace all variables in `n` with ground values of the type.
+   *
+   * @param n A term containing variables
+   * @return `n` with all vars in `n` replaced with ground values
+   */
+  Node mkGround(Node n) const;
+
+  /**
+   * A notification that s is equal to n * { vars -> subs }. This function
```

Comment:
Can you say when in the algorithm this is called?

Reply:
Done.

Reply:
Done.